### PR TITLE
Clarify dependency policy of Sling bundles

### PR DESCRIPTION
**Pull request by kwin**
This clarification was requested in the context of https://lists.apache.org/thread/wbcx13qlt5frkoz1p0jslwf0znqkb6h2.

### CODE REVIEW

**Review by enapps-enorman — CHANGES_REQUESTED** [review]
The policy of depending on the lowest possible version is unnecessarily rigid.  It creates a bunch of confusion for everyone who is using tooling to scan for security problems.

I hope you would reconsider changing this to be the lowest possible version that does not have known security vulnerabilities.  

For me, that is the practical compromise solution that continues the goal of supporting a wide range of environments while also signaling that we care about (and are paying attention to) security vulnerabilities.

**Comment by kwin** [comment]
For me this clarification is primarily about setting the right expectation for our users. I don't think we would be able to timely do new releases in case e.g. a vulnerability in Commons IO or Commons Lang is detected which is used in hundreds of our bundles...

**Comment by enapps-enorman** [comment]
> e.g. a vulnerability in Commons IO or Commons Lang

commons-io has not had a known security vulnerability since version 2.6 (released on May 27, 2020)
commons-lang3 has not had any known security vulnerabilities at all

My impression is that people are using these guidelines as justification for not doing any work at all, or to prevent others from evolving the code to more modern dependencies.  

For example, for SLING-12184 where you raised an objection about bumping up to a 5 year old newer dependency and remain on a 12 year old dependency!

**Comment by kwin** [comment]
Raising a dependency is totally fine if your require certain features or new API only available in a newer version. Just raising the dependencies to get rid of vulnerabilities is not useful!

**Comment by enapps-enorman** [comment]
> Just raising the dependencies to get rid of vulnerabilities is not useful!

That is just not true at all.  There is value in not getting a bunch of false positives from security scanning tools and having to explain the reason every time!  

Bumping the dependencies to a newer version is trivial and only has to be done once in a while and the dependabot does most of the work for you anyways.

**Comment by kwin** [comment]
Obviously we don't agree here, so I would appreciate to hear other opinions @cziegeler and @rombert...

**Comment by reschke** [comment]
Throwing in my 2 cents as innocent bystander:

- as a maintainer of an OSS library, I feel offended when people keep referencing an EOLd version of the library for no good reason
- furthermore, if allowing to run/deploy with an ancient version is an excuse not to *also* test with a current version, something is very wrong (so the both "oldest compatible" and "latest and greatest" should be tested with)
- the latter gets indeed messy when there are many permutations to test with

So by all means drop support for versions that have been officially EOLd.

**Comment by cziegeler** [comment]
We all (open source projects, companies) get swamped with automatically generated security scan reports - these reports have no idea about Maven scopes, modularity, OSGi, you name it. The only way to get ahead of that flood is to update our dependencies to versions without known security vulnerabilities. If we don't do this, our modules look like they have security vulnerabilities and we do not care. It is pretty hard and time consuming to fight this with explanations. The only thing that works are actions - and that is updating the dependencies.

But we are also not doing a good service for our users if we allow them to use insecure dependencies. If we update our dependencies, we force our users to a more secure world.

However, updating a dependency in a pom to a newer version is usually a very cheap change. Releasing the module with the updated dependency is not that cheap anymore. At the same time, we most likely will always be behind with updating - or simply overlook a report for some time. Which means, our users still have to look for security vulnerabilities themselves and update their projects.

Taking all of this together, I think it is good to set a clear expectation to our users that they have to take care of their systems and update dependencies as required. That is the nature of a modular system.
On the other end, we should update dependencies with known vulnerabilities to a version that has no known vulnerabilities. But we need to clearly state that this only makes security scans happy. It does not fix anything anywhere.

**Comment by kwin** [comment]
So which changes of this PR do you guys consider useful or would you rather leave the site as is?

**Comment by cziegeler** [comment]
Very good points. While it might be hard to get all that data, I would expect that installations out there are fairly up to date with security updates.
I would also like to add the question "how painful is it to update the required dependencies?" Since we started with the policy of lowest possible dependencies, a lot has changed. Especially tooling is far better these days. If someone wants to update a bundle, tooling will tell you before the deployment whether that bundle will resolve or needs additional things. Some tooling will also help you in finding these things. Based on that, it is probably less of a deal to update a Sling module together with it dependencies - of course as long as this does not require any code changes.

File: src/main/jbake/content/documentation/development/dependency-management.md [hunk 1]
```diff
@@ -15,22 +15,21 @@ Maven provides projects with a nice feature called dependency management. In Sli
 
 After working with this some time and trying to upgrade various dependencies we came to the conclusion, that using Maven dependency management is not going to work out in the Sling scenario.
 
-Why ? Maven's dependency management is aimed at traditional applicaitons, which are glued together statically during the build process. For this environment, dependency management is a great thing, since it guarantees a consistent application setup.
+Why? Maven's dependency management is aimed at traditional applications, which are glued together statically during the build process. For this environment, dependency management is a great thing, since it guarantees a consistent application setup.
```
Comment (cziegeler) [inline src/main/jbake/content/documentation/development/dependency-management.md:19]: This change looks good

File: src/main/jbake/content/documentation/development/dependency-management.md [hunk 2]
```diff
@@ -15,22 +15,21 @@ Maven provides projects with a nice feature called dependency management. In Sli
...
 
-Why's that ? The Maven Bundle Plugin constructs the bundle manifest and will generally automatically create the Import-Package header. If the providing library (from Maven's dependency list) has `Export-Package` headers with version numbers, the Maven Bundle Plugin will insert the respective version numbers for the `Import-Package` header. This makes perfect sense, because it is expected, that the artifact required at least the given package version.
+Why's that? The Maven Bundle Plugin (or rather the underlying [Bnd library](https://bnd.bndtools.org/) constructs the bundle manifest and will generally automatically create the Import-Package header. If the providing library (from Maven's dependency list) has `Export-Package` headers with version numbers, the Maven Bundle Plugin will insert the respective version numbers for the `Import-Package` header. This makes perfect sense, because it is expected, that the artifact required **at least** the given package version.
```
Comment (cziegeler) [inline src/main/jbake/content/documentation/development/dependency-management.md:22]: This change looks good

File: src/main/jbake/content/documentation/development/dependency-management.md [hunk 3]
```diff
@@ -15,22 +15,21 @@ Maven provides projects with a nice feature called dependency management. In Sli
...
 
 When using Maven dependency management, upgrading any dependencies in the parent POM may automatically increase the version numbers in the `Import-Package` headers and hence may cause any such bundle to fail resolution if deployed - even though the bundle did not change and does not really require a new version of the dependency.
 
 So, in the case of OSGi deployment, Maven's dependency management actually interferes with the OSGi framework dependency management.
 
-As a consequence, I suggest we drop dependency management in the parent POM (almost) completely and state the following.
+As a consequence, we dropped dependency management in the parent POM (almost) completely and state the following.
```
Comment (cziegeler) [inline src/main/jbake/content/documentation/development/dependency-management.md:28]: This change looks good

File: src/main/jbake/content/documentation/development/dependency-management.md [hunk 4]
```diff
@@ -39,28 +38,21 @@ The parent POM only does dependency management for build time dependencies and a
...
-| biz.aQute | bndlib | 1.50.0 | provided |
-| junit | junit | 4.11 | test |
-| org.jmock | jmock-junit4 | 2.5.1 | test |
-| org.slf4j | slf4j-simple | 1.5.2 | test |
+* OSGi spec chapter dependencies (in sling-bundle-parent)
+* Some testing libraries (like JUnit)
+* Some common JSRs (Servlet, JCR, JSR-330)
+* SLF4J 
 
+For details refer to the `pom.xml` of [sling-bundle-parent](https://github.com/apache/sling-parent/blob/master/sling-bundle-parent/pom.xml) and [sling](https://github.com/apache/sling-parent/blob/master/sling-parent/pom.xml).
```
Comment (cziegeler) [inline src/main/jbake/content/documentation/development/dependency-management.md:46]: These changes look good

File: src/main/jbake/content/documentation/development/dependency-management.md [hunk 5]
```diff
@@ -39,28 +38,21 @@ The parent POM only does dependency management for build time dependencies and a
...
 
 All dependencies per module are fully described in terms of version, scope, and classifier by the respective project.
 
-The version of the module dependency should be selected according to the following rule: The lowest version providing the functionality required by the module (or bundle). By required functionality we bascially mean provided API.
+The version of the module dependency should be selected according to the following rule: **The lowest version providing the functionality required by the module (or bundle)**. By required functionality we basically mean provided API.
```
Comment (cziegeler) [inline src/main/jbake/content/documentation/development/dependency-management.md:50]: This change looks good

File: src/main/jbake/content/documentation/development/dependency-management.md [hunk 6]
```diff
@@ -39,28 +38,21 @@ The parent POM only does dependency management for build time dependencies and a
...
 
 Generally there is a constant flow of releases of dependent libraries. In general this should not cause the dependency version number of a using module to be increased. There is one exception though: If the fixed library version contains a bug fix, which has an influence on the operation of the module, an increase in the version number is indicated and should also be applied.
 
 
 ## References
 
-* [Dependency Management](http://markmail.org/message/5qpmsukdk4mdacdy) -- Discussion thread about reducing Maven Dependency Management
+* [Dependency Management](https://lists.apache.org/thread/gbx1t3kfcvqkoljb8mk7ymow94kn2m2o) -- Discussion thread about reducing Maven Dependency Management
 * [SLING-811](https://issues.apache.org/jira/browse/SLING-811) -- The actual issue governing the changes to the project descriptors
```
Comment (cziegeler) [inline src/main/jbake/content/documentation/development/dependency-management.md:58]: This change looks good

File: src/main/jbake/content/project-information/security.md [hunk 7]
```diff
@@ -12,6 +12,10 @@ We strongly encourage folks to report such problems to our private security mail
...
 
+In Sling OSGi bundles we have long had a policy of depending on the lowest possible version of a library/API, to ensure that our bundles are deployable in the widest possible range of environments. Therefore the responsibility of
+ensuring that the environment is secure lies with the assembler and/or deployer of the application, which should make sure that the OSGi bundles they deploy are secure. As such, **we don't consider vulnerable dependencies of our bundles as security issues** by themselves. Usually the dependencies used by Sling [are semantically versioned](https://docs.osgi.org/whitepaper/semantic-versioning/index.html) and therefore security related version updates are fully binary backwards-compatible.
+Further detail and some exceptions from that policy are outlined in [our wiki](https://cwiki.apache.org/confluence/display/SLING/Dependabot).
+
```
Comment (cziegeler) [inline src/main/jbake/content/project-information/security.md:18]: I think this change is ok as well as it does not say anything about *not* updating a dependency. It just provides additional important context

**Comment by kwin** [comment]
@enapps-enorman Can you propose changes for sections you don't agree with or are you fine with the remarks from @cziegeler?

**Comment by enapps-enorman** [comment]
@kwin 
> Can you propose changes for sections you don't agree with or are you fine with the remarks from @cziegeler?

The following is the only statement that I would like to see more clarity on since the "lowest version" doesn't consider known security vulnerabilities when choosing the version to depend on.
> The version of the module dependency should be selected according to the following rule: **The lowest version providing the functionality required by the module (or bundle)**. By required functionality we basically mean provided API.

For me, the rules for dependency version should prefer the oldest version that doesn't contain any known security vulnerability with guidelines like this:
1.  When the generated "Import-Package" clause does not materially change then it should be preferred to bump the dependency to this version.  This would make the security scanning tools not flag this as problem and should not alter the binary compatibility at all.
2. When the generated "Import-Package" clause does materially change, then bump the "minor segment" of the bundle version number and continue further development with this as the new base version number.  Changing the "minor segment" of the bundle version leaves that previous minor bundle version namespace available for any fixes that need to be backported and released on the older branch.  This should make it possible to support consumers who can't upgrade to the new version.  This backport scenario would probably be rare so I would treat it as an "on demand" situation.

I would imagine that some of this decision could be assisted with a bnd plugin that could do some calculations against the baseline version of the bundle.  Maybe some codified tooling that enforces the above and the other bundle version rules from https://sling.apache.org/documentation/development/version-policy.html#evolution-of-bundle-versions

**Comment by cziegeler** [comment]
For me the key word is "should" - it doesn't prevent us from updating a dependency for whatever reason, being it a security bug or a regression or whatever. Once we try to come up with a policy that fits all use cases, we most likely will spent a lot of time on it, just to find out that we missed something. And even if we think we can do this perfect policy are we seriously thinking about having a policy police?

**Comment by enapps-enorman** [comment]
> For me the key word is "should" - it doesn't prevent us from updating a dependency for whatever reason, being it a security bug or a regression or whatever. Once we try to come up with a policy that fits all use cases, we most likely will spent a lot of time on it, just to find out that we missed something. And even if we think we can do this perfect policy are we seriously thinking about having a policy police?

Well, if you read my suggestion again you would see that it says "should" as well.  It just tries to define some best practice guidance.  If that is too complex for you, then go ahead and leave it ambiguous and continue to be afraid of changing anything.

**Comment by cziegeler** [comment]
If we all agree that this is a "should" then I'm totally happy.

**Comment by kwin** [comment]
@enapps-enorman I added your remark in https://github.com/apache/sling-site/pull/149/commits/86bc15dcd68ce98a4cd91b1c9089d4e3d159b79b. Feel free to enhance in a follow-up PR. My main point is for consumers (i.e. they must not expect that all Maven dependencies are not having known vulnerabilities).

**Pull request merged**